rats-devtools/src/python/rats/devtools/_groups.py

```diff
@@ -0,0 +1,77 @@
+from abc import abstractmethod
+from collections.abc import Callable, Iterator
+from typing import Protocol
+
+import click
+
+from rats import apps
+
+from ._plugins import PluginRunner
+
+
+class CommandGroupPlugin(Protocol):
+    @abstractmethod
+    def on_group_registration(self, group: click.Group) -> None:
+        pass
+
+
+class CommandGroup(apps.Executable):
+    _plugins: PluginRunner[CommandGroupPlugin]
+
+    def __init__(self, plugins: PluginRunner[CommandGroupPlugin]) -> None:
+        self._plugins = plugins
+
+    def execute(self) -> None:
+        cli = click.Group()
+        self._plugins.apply(lambda plugin: plugin.on_group_registration(cli))
+        cli()
+
+
+class CommandProvider:
+    _command_names: frozenset[str]
+    _service_mapper: Callable[[str], apps.ServiceId[click.Command]]
+    _container: apps.Container
+
+    def __init__(
+        self,
+        command_names: frozenset[str],
+        service_mapper: Callable[[str], apps.ServiceId[click.Command]],
+        container: apps.Container,
+    ) -> None:
```

Comment:
1. What's the advantage of taking `command_names` and `service_mapper` over taking a `command_name_to_id: Mapping[str, apps.ServiceId[click.Command]`?
I think the mapping is simpler and having to provide the set of command names negates the advantage of the a callable.
2. Why ask for service ids and a container rather than for service providers?

In summary:
```suggestion
    def __init__(
        self,
        commands: Mapping[str, Callable[[], click.Command]],
    ) -> None:
```

Reply:
i guess one issue i have is with the lack of an immutable mapping type. my version only has immutable arguments but would you recomment using `MappingProxyType`?  i don't think i have strong opinions so let me try switching it up

Reply:
AFAIU Mapping is read only.  At least that's how I read (the very sparse documentation here) https://docs.python.org/3/library/collections.abc.html#collections.abc.Mapping.

Reply:
See the entries for Mapping and MutableMapping in the table here: https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes

Reply:
oh you're right, that's not the page i landed on with my search. just pushed a version with your api